framework/src/main/java/org/checkerframework/framework/flow/CFAbstractStore.java

```diff
@@ -199,54 +199,73 @@ public void updateForMethodCall(
         if (!(analysis.checker.hasOption("assumeSideEffectFree")
                 || analysis.checker.hasOption("assumePure")
                 || isSideEffectFree(atypeFactory, method))) {
-            // update field values
-            Map<FieldAccess, V> newFieldValues = new HashMap<>();
-            for (Map.Entry<FieldAccess, V> e : fieldValues.entrySet()) {
-                FieldAccess fieldAccess = e.getKey();
-                V otherVal = e.getValue();
 
-                // case 3: the field has a monotonic annotation
-                if (!((GenericAnnotatedTypeFactory<?, ?, ?, ?>) atypeFactory)
-                        .getSupportedMonotonicTypeQualifiers()
-                        .isEmpty()) {
-                    List<Pair<AnnotationMirror, AnnotationMirror>> fieldAnnotations =
-                            atypeFactory.getAnnotationWithMetaAnnotation(
-                                    fieldAccess.getField(), MonotonicQualifier.class);
-                    V newOtherVal = null;
-                    for (Pair<AnnotationMirror, AnnotationMirror> fieldAnnotation :
-                            fieldAnnotations) {
-                        AnnotationMirror monotonicAnnotation = fieldAnnotation.second;
-                        Name annotation =
-                                AnnotationUtils.getElementValueClassName(
-                                        monotonicAnnotation, "value", false);
-                        AnnotationMirror target =
-                                AnnotationBuilder.fromName(
-                                        atypeFactory.getElementUtils(), annotation);
-                        // Make sure the 'target' annotation is present.
-                        if (AnnotationUtils.containsSame(otherVal.getAnnotations(), target)) {
-                            newOtherVal =
-                                    analysis.createSingleAnnotationValue(
-                                                    target, otherVal.getUnderlyingType())
-                                            .mostSpecific(newOtherVal, null);
+            // update local variables
+            // TODO: Also remove if any element/argument to the annotation is not
+            // isUnmodifiableByOtherCode.  Example: @KeyFor("valueThatCanBeMutated").
+            if (analysis.checker.sideEffectsUnrefineAliases) {
+                localVariableValues
+                        .entrySet()
+                        .removeIf(e -> !e.getKey().isUnmodifiableByOtherCode());
+            }
+
+            // update this value
+            if (analysis.checker.sideEffectsUnrefineAliases) {
+                thisValue = null;
+            }
+
+            // update field values
+            if (analysis.checker.sideEffectsUnrefineAliases) {
```

Comment:
Nit: I would probably move move this if/else to a helper method to reduce the indentation.

Reply:
Let's put this off for now, and revisit the indentation issue later.